File: p3d2x3d.c
/*
 * p3d2x3d: This program outputs p3d visualisations in x3d format for the web
 *
 * LT and WvS November 2015
 *
 * To compile:
 * $ g++ -o p3d2x3d p3d2x3d.c -stdlib=libstdc++ `psrchive --cflags --ldflags`
 * *
 */

#include "Pulsar/Archive.h"
#include "Pulsar/Poincare.h"
#include "Pulsar/Integration.h"
#include "Pulsar/Profile.h"
#include <iostream>
#include <fstream>
#include <stdio.h>
#include <stdlib.h>

using namespace std;

/* Function to get min and max value */
void minmax (float& min, float& max, float val)
{
  if (val < min)
    min = val;
  if (val > max)
    max = val;
}

/* Main */
int main (int argc, char** argv)
{
  if (argc < 2) {
    cerr << "Please specify a filename to process!" << endl;
    return -1;
  }

  /* Defaults */  
  int i, isub = 0, ichan = 0;
  float xmin = 0, ymin = 0, zmin = 0;
  float xmax = 0, ymax = 0, zmax = 0;

  /* Instantiate the archive object */
  Reference::To<Pulsar::Archive> archive;

  /* Load the archive file */
  char * infile = argv[1];
  archive = Pulsar::Archive::load(infile);
  
  /* Remove baseline */
  archive->remove_baseline();
  
  /* Perform frequency scrunch to single channel */
  archive->fscrunch();
  
  /* Perform time scrunch to single subintegration*/
  archive->tscrunch();
  
  /* Check file contains polarisation information */
  if (archive->get_npol() != 4)
  {
    cerr << "WARNING: data file does not contain full polarisation data" << endl;
    return -1;
  }

  /* Convert to Stokes parameters */
  archive->convert_state(Signal::Stokes);
  
  /* Instantiate the subintegration object */
  Pulsar::Integration* integration = archive->get_Integration(isub);

  /* Get all Stokes parameters */
  Pulsar::Profile* Q = integration->get_Profile (1, ichan);
  Pulsar::Profile* U = integration->get_Profile (2, ichan);
  Pulsar::Profile* V = integration->get_Profile (3, ichan);

  /* Get the amplitudes for each Stokes parameter */
  float* Qamps = Q->get_amps();
  float* Uamps = U->get_amps();
  float* Vamps = V->get_amps();
  
  /* Get centre frequency */
  double cfreq = archive->get_centre_frequency();

  /* Get source name */
  string strjname = archive->get_source(); 

  /* Get number of bins */
  int N = archive->get_nbin();
  
  /* Define arrays */  
  float x[N], y[N], z[N];
  
  /* Define and open output file */
  ofstream outfile;
  char outfilename[80];
  string strfilename = string("J") + strjname + string(".x3d");
  strfilename.copy(outfilename,80);
  outfile.open(outfilename);
 
  /* Print file details to stdout */
  cout << "-- p3d2x3d --" << endl;
  cout << "Input file name: " << infile << endl;
  cout << "X3D output file name: " << strfilename << endl;
  cout << "Source: " << strjname << endl;
  cout << "Centre frequency: " << cfreq << endl;
  cout << "NBIN: " << N << endl;
  cout << "NCHAN: " << ichan << endl;
  cout << "NSUB: " << isub << endl;
  
  /* Write xml header */
  outfile << "<X3D profile=\"Immersive\" version=\"3.2\" width=\"400px\" height=\"400px\">" << endl;
  
  /* Output the xml for the x3d dom */
  outfile << "<Scene>" << endl;
  outfile << "<Shape>" << endl;
  outfile << "<Sphere radius=\"4.7\"/>" << endl;
  outfile << "<Appearance>" << endl;
  outfile << "<Material transparency=\"0.89\"/>" << endl;
  outfile << "</Appearance>" << endl;
  outfile << "</Shape>" << endl;
  outfile << "<Group>" << endl;
  /* Y-axis arrow and label */ 
  outfile << "<Group DEF=\"ArrowBlue\">" << endl;
  outfile << "<Shape>" << endl;
  outfile << "<Cylinder DEF=\"ArrowCylinder\" radius=\".005\" top=\"false\"/>" << endl;
  outfile << "<Appearance DEF=\"Blue\">" << endl;
  outfile << "<Material diffuseColor=\"0 0 1\" emissiveColor=\"0 0 1\"/>" << endl;
  outfile << "</Appearance>" << endl;
  outfile << "</Shape>" << endl; 
  outfile << "<Transform translation=\"0 1 0\">" << endl;
  outfile << "<Shape>" << endl;
  outfile << "<Cone DEF=\"ArrowCone\" bottomRadius=\".05\" height=\".1\"/>" << endl;
  outfile << "<Appearance USE=\"Blue\"/>" << endl;
  outfile << "</Shape>" << endl;
  outfile << "</Transform>" << endl;
  outfile << "</Group>" << endl;
  outfile << "<Transform translation=\"0 0.8 0\">" << endl;
  outfile << "<Billboard>" << endl;
  outfile << "<Shape>" << endl;
  outfile << "<Appearance USE=\"Blue\"/>" << endl;
  outfile << "<Text string='\"V\"'>" << endl;
  outfile << "<FontStyle DEF=\"LABEL_FONT\" family='\"SANS\"' justify='\"MIDDLE\" \"MIDDLE\"' size=\".2\"/>" << endl;
  outfile << "</Text>" << endl;
  outfile << "</Shape>" << endl;
  outfile << "</Billboard>" << endl;
  outfile << "</Transform>" << endl;
  outfile << "</Group>" << endl;
  outfile << "<Transform rotation=\"0 0 1 -1.57079\">" << endl;
  /* X-axis arrow and label */
  outfile << "<Group>" << endl;
  outfile << "<Group DEF=\"ArrowGreen\">" << endl;
  outfile << "<Shape>" << endl;
  outfile << "<Cylinder USE=\"ArrowCylinder\"/>" << endl; 
  outfile << "<Appearance DEF=\"Green\">" << endl;
  outfile << "<Material diffuseColor=\"0 1 0\" emissiveColor=\"0 1 0\"/>" << endl;
  outfile << "</Appearance>" << endl;
  outfile << "</Shape>" << endl;
  outfile << "<Transform translation=\"0 1 0\">" << endl;
  outfile << "<Shape>" << endl;
  outfile << "<Cone USE=\"ArrowCone\"/>" << endl; 
  outfile << "<Appearance USE=\"Green\"/>" << endl;
  outfile << "</Shape>" << endl;
  outfile << "</Transform>" << endl;
  outfile << "</Group>" << endl;
  outfile << "<Transform rotation=\"0 0 1 1.57079\" translation=\"0.3 1.3 0\">" << endl;
  outfile << "<Billboard>" << endl;
  outfile << "<Shape>" << endl;
  outfile << "<Appearance USE=\"Green\"/>" << endl; 
  outfile << "<Text string='\"U\"'>" << endl;
  outfile << "<FontStyle USE=\"LABEL_FONT\"/>" << endl;
  outfile << "</Text>" << endl;
  outfile << "</Shape>" << endl;
  outfile << "</Billboard>" << endl;
  outfile << "</Transform>" << endl;
  outfile << "</Group>" << endl;
  outfile << "</Transform>" << endl;
  outfile << "<Transform rotation=\"1 0 0 1.57079\">" << endl;
  outfile << "<Group>" << endl;
  /* Z-axis arrow and label */
  outfile << "<Group DEF=\"ArrowRed\">" << endl;
  outfile << "<Shape>" << endl;
  outfile << "<Cylinder USE=\"ArrowCylinder\"/>" << endl;
  outfile << "<Appearance DEF=\"Red\">" << endl;
  outfile << "<Material diffuseColor=\"1 0 0\" emissiveColor=\"1 0 0\"/>" << endl;
  outfile << "</Appearance>" << endl;
  outfile << "</Shape>" << endl;
  outfile << "<Transform translation=\"0 1 0\">" << endl;
  outfile << "<Shape>" << endl;
  outfile << "<Cone USE=\"ArrowCone\"/>" << endl;
  outfile << "<Appearance USE=\"Red\"/>" << endl;
  outfile << "</Shape>" << endl;
  outfile << "</Transform>" << endl;
  outfile << "</Group>" << endl;
  outfile << "<Transform rotation=\"1 0 0 -1.57079\" translation=\"0 1.3 0.3\">" << endl;
  outfile << "<Billboard>" << endl;
  outfile << "<Shape>" << endl;
  outfile << "<Appearance USE=\"Red\"/>" << endl;
  outfile << "<Text string='\"Q\"'>" << endl;
  outfile << "<FontStyle USE=\"LABEL_FONT\"/>" << endl;
  outfile << "</Text>" << endl;
  outfile << "</Shape>" << endl;
  outfile << "</Billboard>" << endl;
  outfile << "</Transform>" << endl;
  outfile << "</Group>" << endl;
  outfile << "</Transform>" << endl;
  outfile << "<Shape>" << endl;
  outfile << "<Appearance>" << endl;
  outfile << "<Material emissiveColor=\"1 1 1\"/>" << endl;
  outfile << "</Appearance>" << endl;
  outfile << "<LineSet vertexCount=\"" << N << "\" containerField=\"geometry\">" << endl;
  outfile << "<Coordinate point=\"";

  /* Loop over the bins */
  for (unsigned ibin=0; ibin < N; ibin++)
    {
      x[ibin] = Qamps[ibin]*10;
      y[ibin] = Uamps[ibin]*10;
      z[ibin] = Vamps[ibin]*10;
      outfile << x[ibin] << " " << y[ibin] << " " << z[ibin] << ",";  
      minmax (xmin, xmax, x[ibin]);
      minmax (ymin, ymax, y[ibin]);
      minmax (zmin, zmax, z[ibin]);
    }

  /* Complete the x3d dom */
  outfile << "\">" << endl;
  outfile << "</Coordinate>" << endl;
  outfile << "</LineSet>" << endl;
  outfile << "</Shape>" << endl;
  outfile << "</Scene>" << endl;
  outfile << "</X3D>" << endl;

  /* Close output file */
  outfile.close();

  return 0;
}

File: p3d2x3d.py
#!/usr/bin/env python

"""p3d2x3d.py

This program writes HTML and X3D format files of PSRCHIVE 'p3d-style'
polarisation data for 3D visualsation in X3D enabled web browsers 
and X3D viewers

Copyright (C) 2016  Lawrence Toomey

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""

import psrchive
import argparse
from lxml import etree as ET

isub = 0
ibin = 0
ichan = 0
nbin = 0
str_points = ''
red_rgb = '1 0 0'
green_rgb = '0 1 0'
blue_rgb = '0 0 1'
white_rgb = '1 1 1' 

def draw(root_element, shape_type, emissive_colour, diffuse_colour, \
         transparency, height, radius):
    """Function to render a shape in the X3DOM
    
    Renders a given shape type in a given parent element of the X3DOM

    Arguments:
        root_element: Parent element in which to place 
                      the shape sub element
        shape_type: Shape type [options: Sphere|Cylinder|Cone]
        emissive_colour: Emissive colour [R G B]
        diffuse_colour: Diffuse colour [R G B]
        transparency: Amount of tranparency [0(None):1(transparent)]
        height: Element height
        radius: Element radius

    Returns:
        An X3D element containing the required shape, as an xml string
    """

    s_elem = ET.SubElement(root_element, 'Shape')
    
    if shape_type == 'Sphere':
        s_sub_elem = ET.SubElement(s_elem, shape_type, radius=radius)
    elif shape_type == 'Cylinder':
        s_sub_elem = ET.SubElement(s_elem, shape_type, radius=radius, \
                                   top='TRUE')
    elif shape_type == 'Cone':
        s_sub_elem = ET.SubElement(s_elem, shape_type, bottomRadius=radius, \
                                   height=height)
     
    s_app = ET.SubElement(s_elem, 'Appearance')
    s_material = ET.SubElement(s_app, 'Material', \
                               emissiveColor=emissive_colour, \
                               diffuseColor=diffuse_colour, \
                               transparency=transparency)
    
    return ET.tostring(s_elem)

def label(root_element, shape_type, emissive_colour, diffuse_colour, \
          transparency, string, font_type, font_size, justify):
    """Function to add a label to an element in the X3DOM
    
    Adds a text label to a given parent element in the X3DOM

    Arguments:
        root_element: Parent element in which to place 
                      the label sub element
        shape_type: Shape type [options: Text]
        emissive_colour: Emissive colour [R G B]
        diffuse_colour: Diffuse colour [R G B]
        transparency: Amount of tranparency [0(None):1(transparent)]
        string: Label text string
        font_type: Font family
        font_size: Font size
        justify: Label position

    Returns:
        An X3D element containing a label, as an xml string            
    """

    l_elem = ET.SubElement(root_element, 'Shape')
    l_sub_elem = ET.SubElement(l_elem, 'Text', string=string)
    l_sub_elem_font = ET.SubElement(l_sub_elem, 'FontStyle', \
                                    family=font_type, \
                                    size=font_size, \
                                    justify=justify)
    l_app = ET.SubElement(l_elem, 'Appearance')
    l_material = ET.SubElement(l_app, 'Material', \
                               emissiveColor=emissive_colour, \
                               diffuseColor=diffuse_colour, \
                               transparency=transparency)

    return ET.tostring(l_elem)

"""Parse command-line arguments"""
parser = argparse.ArgumentParser(usage='p3d2x3d.py file\n'
         '\nThis program writes HTML and X3D format files of PSRCHIVE p3d-style\n'
         'polarisation data for 3D visualsation in X3D enabled web browsers\n' 
         'and X3D viewers\n'
         '\nCopyright (C) 2016  Lawrence Toomey\n'
         'This program comes with ABSOLUTELY NO WARRANTY.\n'
         'This is free software, and you are welcome to redistribute it\n'
         'under certain conditions.\n')
parser.add_argument('file', type=argparse.FileType('r'))
args = parser.parse_args()

"""Instantiate the archive object and load the input file"""
in_file = args.file.name
try:
    ar = psrchive.Archive_load(in_file)
except:
    print 'ERROR: problem loading file %s. Exiting.' %in_file

"""Check the file contains full polarisation data"""
npol = ar.get_npol()
if npol != 4:
    print 'ERROR: file %s does not contain' \
          ' full polarisation data. Exiting.' %in_file
    quit()

"""Remove the baseline"""
ar.remove_baseline()

"""Frequency scrunch to a single channel"""
ar.fscrunch()
    
"""Time scrunch to a single subintegration"""
ar.tscrunch()
	
"""Convert to Stokes parameters"""
ar.convert_state('Stokes')
    
"""Instantiate the subintegration object"""
subint = ar.get_Integration(isub)

"""Get the Stokes Q, U and V parameters"""
Q = subint.get_Profile(1, ichan)
U = subint.get_Profile(2, ichan)
V = subint.get_Profile(3, ichan)

"""Get the amplitudes for each Stokes parameter"""
Qamps = Q.get_amps()
Uamps = U.get_amps()
Vamps = V.get_amps()
			  
"""Get the centre frequency"""
cfreq = ar.get_centre_frequency()

"""Get the source name"""
src = ar.get_source()

"""Get the number of bins"""
nbin = ar.get_nbin()

"""Define the output file names and open for writing"""
x3d_file_name = src + '.x3d'
html_file_name = src + '.html'
x3d_file = open(x3d_file_name, 'w')
html_file = open(html_file_name, 'w')

"""Print file details to stdout"""
print '\n-- p3d2x3d --'
print 'Input file name:        %s' %in_file
print 'Source:                 %s' %src
print 'Centre frequency:       %f' %cfreq
print 'No. of polarisations:   %d' %npol
print 'No. of bins:            %d' %nbin

"""Build the X3DOM"""
root = ET.Element('X3D', profile='Immersive', version='3.2')
scene = ET.SubElement(root, 'Scene')

"""Add the title"""
title_label_transform = ET.SubElement(scene, 'Transform', \
                                      translation='0 2 0')
title_billboard = ET.SubElement(title_label_transform, 'Billboard')
label(title_billboard, 'Text', white_rgb, white_rgb, '0', \
      '"'  + src + ' ' + str(cfreq) + 'MHz"', '"SANS"', '0.2', \
      '"MIDDLE" "MIDDLE"')

"""Draw the cylinder for the Stokes U (X-axis) vector"""
x_transform = ET.SubElement(scene, 'Transform', rotation='0 0 1 -1.57079')
x_group = ET.SubElement(x_transform, 'Group')
x_group_def = ET.SubElement(x_group, 'Group')
draw(x_group_def, 'Cylinder', green_rgb, green_rgb, '0', '0', '0.005')

"""Draw the cone for the Stokes U vector"""
x_cone_transform = ET.SubElement(x_group_def, 'Transform', translation='0 1 0')
draw(x_cone_transform, 'Cone', green_rgb, green_rgb, '0', '0.1', '0.05')

"""Add the label to the Stokes U vector"""
x_label_transform = ET.SubElement(x_group, 'Transform', \
                                  rotation='0 0 1 1.57079', \
                                  translation='0.3 1.3 0')
x_billboard = ET.SubElement(x_label_transform, 'Billboard')
label(x_billboard, 'Text', green_rgb, green_rgb, '0', '"U"', '"SANS"', '0.2', \
      '"MIDDLE" "MIDDLE"')

"""Draw the cylinder for the Stokes V (Y-axis) vector"""
y_group = ET.SubElement(scene, 'Group')
y_group_def = ET.SubElement(y_group, 'Group')
draw(y_group_def, 'Cylinder', blue_rgb, blue_rgb, '0', '0', '0.005')

"""Draw the cone for the Stokes V vector"""
y_cone_transform = ET.SubElement(y_group_def, 'Transform', translation='0 1 0')
draw(y_cone_transform, 'Cone', blue_rgb, blue_rgb, '0', '0.1', '0.05')

"""Add the label to the Stokes V vector"""
y_label_transform = ET.SubElement(y_group, 'Transform', \
                                  translation='0 1.2 0')
y_billboard = ET.SubElement(y_label_transform, 'Billboard')
label(y_billboard, 'Text', blue_rgb, blue_rgb, '0', '"V"', '"SANS"', '0.2', \
      '"MIDDLE" "MIDDLE"')

"""Draw the cylinder for the Stokes Q (Z-axis) vector"""
z_transform = ET.SubElement(scene, 'Transform', rotation='1 0 0 1.57079')
z_group = ET.SubElement(z_transform, 'Group')
z_group_def = ET.SubElement(z_group, 'Group')
draw(z_group_def, 'Cylinder', red_rgb, red_rgb, '0', '0', '0.005')

"""Draw the cone for the Stokes Q vector"""
z_cone_transform = ET.SubElement(z_group_def, 'Transform', translation='0 1 0')
draw(z_cone_transform, 'Cone', red_rgb, red_rgb, '0', '0.1', '0.05')

"""Add the label to the Stokes Q vector"""
z_label_transform = ET.SubElement(z_group, 'Transform', \
                                  rotation='1 0 0 -1.57079', \
                                  translation='0 1.3 0.3')
z_billboard = ET.SubElement(z_label_transform, 'Billboard')
label(z_billboard, 'Text', red_rgb, red_rgb, '0', '"Q"', '"SANS"', '0.2', \
      '"MIDDLE" "MIDDLE"')

"""Plot the Stokes Q, U and V polarisation vectors"""
str_nbin = str(nbin)
pol_shape = ET.SubElement(scene, 'Shape')
pol_app = ET.SubElement(pol_shape, 'Appearance')
pol_material = ET.SubElement(pol_app, 'Material', emissiveColor=white_rgb)
line_set = ET.SubElement(pol_shape, 'LineSet', vertexCount=str_nbin, \
                         containerField='geometry')

"""Loop over the bins"""
x = range(nbin)
y = range(nbin)
z = range(nbin)

for ibin in range(nbin):
    x[ibin] = Qamps[ibin]*10
    y[ibin] = Uamps[ibin]*10
    z[ibin] = Vamps[ibin]*10
    x_coord = str(x[ibin])
    y_coord = str(y[ibin])
    z_coord = str(z[ibin])
    str_points = str_points + ' ' + x_coord + \
                 ' ' + y_coord + ' ' + z_coord + ','

coord = ET.SubElement(line_set, 'Coordinate', point=str_points) 

"""Write the X3DOM to the output x3d file"""
tree = ET.ElementTree(root)
tree.write(x3d_file, pretty_print=True, xml_declaration=False)

"""Write the HTML markup to the output html file"""
content = """<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv='Content-Type' content='text/html;charset=utf-8' />
    <title>""" + src + """</title>
    <script type='text/javascript' src='http://www.x3dom.org/download/x3dom.js'></script>
    <style>
      body {background-color: black;}
    </style>
  </head>
  <body>
    <x3d width='100%' height='50%'> 
      <scene>
        <inline url='""" + x3d_file_name + """'></inline> 
      </scene> 
    </x3d> 
  </body>
</html>
"""
html_file.write(content)
    
"""Close the output files"""
print 'X3D output written to:  %s'  %x3d_file_name
print 'HTML output written to: %s\n' %html_file_name
x3d_file.close()
html_file.close()
